File: tags/Platform631/ASCOM.Utilities/ASCOM.Utilities/AllUsersFileSystemProvider.vb
'Implements the filestore mechanic to store ASCOM profiles

'This implementation stores files in the All Users profile area that is accessible to everyone and so 
'creates a "per machine" store rather than a "per user" store.

Imports System.IO
Imports ASCOM.Utilities.Interfaces
Imports System.Security.AccessControl
Imports System.Security.Principal

Friend Class AllUsersFileSystemProvider

    Implements IFileStoreProvider
    Private Const ASCOM_DIRECTORY As String = "\ASCOM"
    Private Const PROFILE_DIRECTORY As String = ASCOM_DIRECTORY & "\Profile" 'Root directory within the supplied file system space

    Private BaseFolder, ASCOMFolder As String

#Region "New"
    Friend Sub New()
        'Find the location of the All Users profile
        ASCOMFolder = System.Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData) & ASCOM_DIRECTORY
        BaseFolder = System.Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData) & PROFILE_DIRECTORY
    End Sub
#End Region

#Region "IFileStore Implementation"
    Friend Sub SetSecurityACLs(ByVal p_tl As TraceLogger) Implements IFileStoreProvider.SetSecurityACLs
        Dim dInfo As New DirectoryInfo(ASCOMFolder) 'Apply to the ASCOM folder itself
        Dim dSecurity As DirectorySecurity

        'PWGS 5.5.2.0 Fix for users security group not being globally usable
        'Build a temp domainSID using the Null SID passed in as a SDDL string. The constructor will 
        'accept the traditional notation or the SDDL notation interchangeably.
        Dim DomainSid As New SecurityIdentifier("S-1-0-0")
        'Create a security Identifier for the BuiltinUsers Group to be passed to the new accessrule
        Dim Ident As New SecurityIdentifier(WellKnownSidType.BuiltinUsersSid, DomainSid)

        p_tl.LogMessage("  SetSecurityACLs", "Retrieving access control")
        dSecurity = dInfo.GetAccessControl

        p_tl.LogMessage("  SetSecurityACLs", "Adding full control access rule")
        dSecurity.AddAccessRule(New FileSystemAccessRule(Ident, FileSystemRights.FullControl, InheritanceFlags.ContainerInherit Or InheritanceFlags.ObjectInherit, PropagationFlags.None, AccessControlType.Allow))

        p_tl.LogMessage("  SetSecurityACLs", "Setting access control")
        dInfo.SetAccessControl(dSecurity)

        p_tl.LogMessage("  SetSecurityACLs", "Successfully set security ACL!")
    End Sub

    Friend ReadOnly Property Exists(ByVal p_FileName As String) As Boolean Implements IFileStoreProvider.Exists
        'Tests whether a file exists and returns a boolean value
        Get
            Try
                If File.Exists(CreatePath(p_FileName)) Then
                    Return True 'This was successful 
                Else
                    Return False 'No exception but no file returned so the file does not exist
                End If
            Catch ex As Exception
                MsgBox("Exists " & ex.ToString)
                Return False 'Exception so file doesn't exist
            End Try
        End Get
    End Property

    Friend Sub CreateDirectory(ByVal p_SubKeyName As String, ByVal p_TL As TraceLogger) Implements IFileStoreProvider.CreateDirectory
        'Creates a directory in the supplied path (p_SubKeyName)
        Try
            p_TL.LogMessage("  CreateDirectory", "Creating directory for: """ & p_SubKeyName & """")
            Directory.CreateDirectory(CreatePath(p_SubKeyName))
            p_TL.LogMessage("  CreateDirectory", "Created directory OK")
        Catch ex As Exception
            p_TL.LogMessage("FileSystem.CreateDirectory", "Exception: " & ex.ToString)
            MsgBox("CreateDirectory Exception: " & ex.ToString)
        End Try
    End Sub

    Friend Sub DeleteDirectory(ByVal p_SubKeyName As String) Implements IFileStoreProvider.DeleteDirectory
        'Deletes a directory specified by p_SubKeyName
        Directory.Delete(CreatePath(p_SubKeyName), True)
    End Sub

    Friend Sub EraseFileStore() Implements IFileStoreProvider.EraseFileStore
        Dim Response As MsgBoxResult
        Response = MsgBox("Are you sure you wish to erase the Utilities profile store?", MsgBoxStyle.OkCancel Or MsgBoxStyle.Critical, "ASCOM.Utilities")
        If Response = MsgBoxResult.Ok Then
            Try : Directory.Delete(BaseFolder, True) : Catch : End Try
        End If
    End Sub

    Friend ReadOnly Property GetDirectoryNames(ByVal p_SubKeyName As String) As String() Implements IFileStoreProvider.GetDirectoryNames
        'Enumerates the sub directories within a directory specified by p_SubKeyName
        'Returns a string array of directory names
        Get
            Dim FullDirs() As String, ct As Integer, RelDirs() As String
            FullDirs = Directory.GetDirectories(CreatePath(p_SubKeyName))
            ct = 0
            For Each FullDir As String In FullDirs
                RelDirs = Split(FullDir, "\")
                FullDirs(ct) = RelDirs(RelDirs.Length - 1)
                ct += 1
            Next
            Return FullDirs
        End Get
    End Property

    Friend ReadOnly Property FullPath(ByVal p_FileName As String) As String Implements IFileStoreProvider.FullPath
        Get
            Return CreatePath(p_FileName)
        End Get
    End Property

    Friend ReadOnly Property BasePath() As String Implements IFileStoreProvider.BasePath
        Get
            Return BaseFolder
        End Get
    End Property

    Friend Sub Rename(ByVal p_CurrentName As String, ByVal p_NewName As String) Implements IFileStoreProvider.Rename
        File.Delete(CreatePath(p_NewName)) 'Make sure the target file doesn't exist
        File.Move(CreatePath(p_CurrentName), CreatePath(p_NewName))
    End Sub

    Friend Sub RenameDirectory(ByVal CurrentName As String, ByVal NewName As String) Implements IFileStoreProvider.RenameDirectory
        Try : Directory.Delete(CreatePath(NewName), True) : Catch : End Try 'Remove driectory if it already exists
        Dim DirInfo As New DirectoryInfo(CreatePath(CurrentName))
        DirInfo.MoveTo(CreatePath(NewName))
    End Sub

#End Region

#Region "Support Code"
    Private Function CreatePath(ByVal p_FileName As String) As String
        If Left(p_FileName, 1) <> "\" Then p_FileName = "\" & p_FileName
        Return BaseFolder & p_FileName
    End Function
#End Region

End Class

File: ASCOM.Utilities/ASCOM.Utilities/AllUsersFileSystemProvider.cs
using System;
// Implements the filestore mechanic to store ASCOM profiles

// This implementation stores files in the All Users profile area that is accessible to everyone and so 
// creates a "per machine" store rather than a "per user" store.

using System.IO;
using System.Security.AccessControl;
using System.Security.Principal;
using ASCOM.Utilities.Interfaces;
using Microsoft.VisualBasic;

namespace ASCOM.Utilities
{


    internal class AllUsersFileSystemProvider : IFileStoreProvider
    {
        private const string ASCOM_DIRECTORY = @"\ASCOM";
        private const string PROFILE_DIRECTORY = ASCOM_DIRECTORY + @"\Profile"; // Root directory within the supplied file system space

        private string BaseFolder, ASCOMFolder;

        #region New
        internal AllUsersFileSystemProvider()
        {
            // Find the location of the All Users profile
            ASCOMFolder = Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData) + ASCOM_DIRECTORY;
            BaseFolder = Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData) + PROFILE_DIRECTORY;
        }
        #endregion

        #region IFileStore Implementation
        internal void SetSecurityACLs(TraceLogger p_tl)
        {
            var dInfo = new DirectoryInfo(ASCOMFolder); // Apply to the ASCOM folder itself
            DirectorySecurity dSecurity;

            // PWGS 5.5.2.0 Fix for users security group not being globally usable
            // Build a temp domainSID using the Null SID passed in as a SDDL string. The constructor will 
            // accept the traditional notation or the SDDL notation interchangeably.
            var DomainSid = new SecurityIdentifier("S-1-0-0");
            // Create a security Identifier for the BuiltinUsers Group to be passed to the new accessrule
            var Ident = new SecurityIdentifier(WellKnownSidType.BuiltinUsersSid, DomainSid);

            p_tl.LogMessage("  SetSecurityACLs", "Retrieving access control");
            dSecurity = dInfo.GetAccessControl();

            p_tl.LogMessage("  SetSecurityACLs", "Adding full control access rule");
            dSecurity.AddAccessRule(new FileSystemAccessRule(Ident, FileSystemRights.FullControl, InheritanceFlags.ContainerInherit | InheritanceFlags.ObjectInherit, PropagationFlags.None, AccessControlType.Allow));

            p_tl.LogMessage("  SetSecurityACLs", "Setting access control");
            dInfo.SetAccessControl(dSecurity);

            p_tl.LogMessage("  SetSecurityACLs", "Successfully set security ACL!");
        }

        void IFileStoreProvider.SetSecurityACLs(TraceLogger p_tl) => SetSecurityACLs(p_tl);

        // Tests whether a file exists and returns a boolean value
        internal bool get_Exists(string p_FileName)
        {
            try
            {
                if (File.Exists(CreatePath(p_FileName)))
                {
                    return true; // This was successful 
                }
                else
                {
                    return false;
                } // No exception but no file returned so the file does not exist
            }
            catch (Exception ex)
            {
                MessageBox.Show("Exists " + ex.ToString());
                return false; // Exception so file doesn't exist
            }
        }

        bool IFileStoreProvider.get_Exists(string p_FileName) => get_Exists(p_FileName);

        internal void CreateDirectory(string p_SubKeyName, TraceLogger p_TL)
        {
            // Creates a directory in the supplied path (p_SubKeyName)
            try
            {
                p_TL.LogMessage("  CreateDirectory", "Creating directory for: \"" + p_SubKeyName + "\"");
                Directory.CreateDirectory(CreatePath(p_SubKeyName));
                p_TL.LogMessage("  CreateDirectory", "Created directory OK");
            }
            catch (Exception ex)
            {
                p_TL.LogMessage("FileSystem.CreateDirectory", "Exception: " + ex.ToString());
                MessageBox.Show("CreateDirectory Exception: " + ex.ToString());
            }
        }

        void IFileStoreProvider.CreateDirectory(string p_SubKeyName, TraceLogger p_TL) => CreateDirectory(p_SubKeyName, p_TL);

        internal void DeleteDirectory(string p_SubKeyName)
        {
            // Deletes a directory specified by p_SubKeyName
            Directory.Delete(CreatePath(p_SubKeyName), true);
        }

        void IFileStoreProvider.DeleteDirectory(string p_SubKeyName) => DeleteDirectory(p_SubKeyName);

        internal void EraseFileStore()
        {
            MsgBoxResult Response;
            Response = MessageBox.Show("Are you sure you wish to erase the Utilities profile store?", MsgBoxStyle.OkCancel | MsgBoxStyle.Critical, "ASCOM.Utilities");
            if (Response == MsgBoxResult.Ok)
            {
                try
                {
                    Directory.Delete(BaseFolder, true);
                }
                catch
                {
                }
            }
        }

        void IFileStoreProvider.EraseFileStore() => EraseFileStore();

        // Enumerates the sub directories within a directory specified by p_SubKeyName
        // Returns a string array of directory names
        internal string[] get_GetDirectoryNames(string p_SubKeyName)
        {
            string[] FullDirs;
            int ct;
            string[] RelDirs;
            FullDirs = Directory.GetDirectories(CreatePath(p_SubKeyName));
            ct = 0;
            foreach (string FullDir in FullDirs)
            {
                RelDirs = Strings.Split(FullDir, @"\");
                FullDirs[ct] = RelDirs[RelDirs.Length - 1];
                ct += 1;
            }
            return FullDirs;
        }

        string[] IFileStoreProvider.get_GetDirectoryNames(string p_SubKeyName) => get_GetDirectoryNames(p_SubKeyName);

        internal string get_FullPath(string p_FileName)
        {
            return CreatePath(p_FileName);
        }

        string IFileStoreProvider.get_FullPath(string p_FileName) => get_FullPath(p_FileName);

        internal string BasePath
        {
            get
            {
                return BaseFolder;
            }
        }

        string IFileStoreProvider.BasePath { get => BasePath; }

        internal void Rename(string p_CurrentName, string p_NewName)
        {
            File.Delete(CreatePath(p_NewName)); // Make sure the target file doesn't exist
            File.Move(CreatePath(p_CurrentName), CreatePath(p_NewName));
        }

        void IFileStoreProvider.Rename(string p_CurrentName, string p_NewName) => Rename(p_CurrentName, p_NewName);

        internal void RenameDirectory(string CurrentName, string NewName)
        {
            try
            {
                Directory.Delete(CreatePath(NewName), true);
            }
            catch
            {
            } // Remove driectory if it already exists
            var DirInfo = new DirectoryInfo(CreatePath(CurrentName));
            DirInfo.MoveTo(CreatePath(NewName));
        }

        void IFileStoreProvider.RenameDirectory(string CurrentName, string NewName) => RenameDirectory(CurrentName, NewName);

        #endregion

        #region Support Code
        private string CreatePath(string p_FileName)
        {
            if (Strings.Left(p_FileName, 1) != @"\")
                p_FileName = @"\" + p_FileName;
            return BaseFolder + p_FileName;
        }
        #endregion

    }
}